extensions/security/runtime/src/main/java/io/quarkus/security/runtime/interceptor/SecurityConstrainer.java

```diff
@@ -0,0 +1,164 @@
+package io.quarkus.security.runtime.interceptor;
+
+import static java.util.Arrays.asList;
+
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Method;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import javax.annotation.security.DenyAll;
+import javax.annotation.security.PermitAll;
+import javax.annotation.security.RolesAllowed;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+import io.quarkus.security.Authenticated;
+import io.quarkus.security.identity.SecurityIdentity;
+import io.quarkus.security.runtime.ForbiddenException;
+import io.quarkus.security.runtime.UnauthorizedException;
+
+/**
+ * @author Michal Szynkiewicz, michal.l.szynkiewicz@gmail.com
+ *         <br>
+ *         Date: 03/10/2019
+ */
+@Singleton
+public class SecurityConstrainer {
+
+    private static final List<Class<? extends Annotation>> SECURITY_ANNOTATIONS = asList(Authenticated.class, DenyAll.class,
+            PermitAll.class, RolesAllowed.class);
+
+    private final Map<Method, Optional<Check>> checkForMethod = new HashMap<>();
+
+    @Inject
+    SecurityIdentity identity;
+
+    public void checkRoles(Method method, Collection<Annotation> interceptorBindings) {
+        getCheck(method, interceptorBindings)
+                .ifPresent(check -> check.apply(identity));
+    }
+
+    private Optional<Check> getCheck(Method method, Collection<Annotation> interceptorBindings) {
+        return checkForMethod.computeIfAbsent(method, m -> determineSecurityCheck(m, interceptorBindings));
+    }
+
+    private Optional<Check> determineSecurityCheck(Method method, Collection<Annotation> interceptorBindings) {
+        Annotation securityAnnotation = determineSecurityAnnotation(method.getDeclaredAnnotations(), method::toString);
+        if (securityAnnotation == null) {
+            Class<?> declaringClass = method.getDeclaringClass();
+            securityAnnotation = determineSecurityAnnotation(declaringClass.getDeclaredAnnotations(),
+                    declaringClass::getCanonicalName);
+        }
+        if (securityAnnotation == null) {
+            securityAnnotation = determineSecurityAnnotationFromBindings(interceptorBindings, method::toString);
+        }
+        return checkForAnnotation(securityAnnotation);
+    }
```

Comment:
So far the codebase has tried to not use lambdas in runtime code unless there is a very good reason to (for performance reasons). Would it be possible to change these instances to non-lambda code?

Reply:
These are method references, are they forbidden too?
They can be translated without problem to classes but we'd lose on readability.
And it's executed only on failure, and only once for a method so not sure if it's worth it.

Reply:
`determineSecurityAnnotationFromBindings` however is full of lambdas :)

Reply:
They use a lot more memory at runtime, and are slower to load which impacts startup time. It is best to avoid them.

Reply:
true about `determineSecurityAnnotationFromBindings` :) I will de-lambdify it